File: Questao 04 - Check/QuickSort.cpp
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

#define TAM 1000

typedef struct
{
    char nomeArquivo[100];
    char nome[100];
    char corDoCabelo[100];
    char corDaPele[100];
    char corDosOlhos[100];
    char anoNascimento[100];
    char genero[100];
    char homeworld[100];
    int altura;
    double peso;
} Personagens;

typedef struct CelulaDupla
{
    Personagens elemento;
    struct CelulaDupla *prox;
    struct CelulaDupla *ant;
} CelulaDupla;

CelulaDupla *primeiro;
CelulaDupla *ultimo;

bool isFim(char *s);
char *substring(char *padrao, char *entrada);
long int indexOf(char *padrao, char *entrada);
char *lerCorDoCabelo(char *line);
char *lerCorDaPele(char *line);
char *lerCorDosOlhos(char *line);
char *lerAnoNascimento(char *line);
char *lerGenero(char *line);
char *lerHomeworld(char *line);
void lerPersonagem();
void printPersonagem(Personagens x);
//void inserirInicio(Personagens x);
void inserirFim(Personagens x);
//void inserirPos(Personagens x, int pos);
//Personagens removerInicio();
//Personagens removerFim();
//Personagens removerPos(int pos);
Personagens personagem;
//Personagens array[TAM];
int compareTo(char *a, char *b, char *c, char *d);
void swap(int a, int b);
void quicksort();
void quicksortRec(int esq, int dir);

int n = 0;

int main()
{
    char palavra[TAM][TAM];

    int i = 0;

    do //preencher a matriz
    {
        fgets(palavra[i], TAM, stdin);
    } while (isFim(palavra[i++]) == false); //testar se eh FIM
    i--;

    for (int contador = 0; contador < i; contador++) //ler os primeiros personagens e inserir na lista
    {
        strcpy(personagem.nomeArquivo, strndup(palavra[contador], strlen(palavra[contador]) - 1));
        lerPersonagem();
        inserirFim(personagem);
    }

    quicksort();

    CelulaDupla *a = primeiro;

    for (; a != ultimo; a = a->prox)
    {
        strcpy(personagem.nomeArquivo, a->elemento.nomeArquivo); //printar os personagem
        printPersonagem(personagem);
    }
    strcpy(personagem.nomeArquivo, a->elemento.nomeArquivo);
    printPersonagem(personagem);
}

CelulaDupla *novaCelulaDupla(Personagens elemento)
{
    CelulaDupla *nova = (CelulaDupla *)malloc(sizeof(CelulaDupla));
    nova->elemento = elemento;
    nova->ant = nova->prox = NULL;
    return nova;
}

CelulaDupla *celulaPos(int x)
{
    CelulaDupla *aux;
    int i = 0;
    for (aux = primeiro; i < x; aux = aux->prox)
    {
        i++;
    }
    return aux;
}

int compareTo(char *a, char *b, char *c, char *d)
{
    int resp = 0;

    int tam = strlen(a);
    if (strlen(b) < tam)
        tam = strlen(b);

    for (int i = 0; i < tam; i++)
    {
        if (a[i] > b[i])
        {
            resp = 1;
            i = tam;
        }

        else if (a[i] < b[i])
        {
            resp = -1;
            i = tam;
        }
    }

    if (resp == 0 && strlen(a) != strlen(b))
    {
        if (strlen(b) > strlen(a))
            resp = -1;
        else
            resp = 1;
    }
    else if (resp == 0)
    {
        int tam = strlen(c);
        if (strlen(d) < tam)
            tam = strlen(d);

        for (int i = 0; i < tam; i++)
        {
            if (c[i] > d[i])
            {
                resp = 1;
                i = tam;
            }

            else if (c[i] < d[i])
            {
                resp = -1;
                i = tam;
            }
        }

        if (resp == 0 && strlen(c) != strlen(d))
        {
            if (strlen(d) > strlen(c))
                resp = -1;
            else
                resp = 1;
        }
    }

    return resp;
}

void swap(int b, int c)
{
    Personagens tmp = celulaPos(b)->elemento;
    celulaPos(b)->elemento = celulaPos(c)->elemento;
    celulaPos(c)->elemento = tmp;
}

void quicksortRec(int esq, int dir)
{
    int i = esq, j = dir;
    Personagens pivo = celulaPos((dir + esq) / 2)->elemento;

    while (i <= j)
    {
        CelulaDupla *aux = celulaPos(i);
        while (compareTo(aux->elemento.corDoCabelo, pivo.corDoCabelo, aux->elemento.nome, pivo.nome) < 0)
        {
            aux = aux->prox;
            i++;
        }

        aux = celulaPos(j);
        while (compareTo(aux->elemento.corDoCabelo, pivo.corDoCabelo, aux->elemento.nome, pivo.nome) > 0)
        {
            aux = aux->ant;
            j--;
        }

        if (i <= j)
        {
            swap(i, j);
            i++;
            j--;
        }
    }

    if (esq < j)
        quicksortRec(esq, j);
    if (i < dir)
        quicksortRec(i, dir);
}

void quicksort()
{
    quicksortRec(0, n - 1);
}

void lerPersonagem()
{
    FILE *arq;
    arq = fopen(personagem.nomeArquivo, "rb");

    char *line = NULL;

    size_t bufsize = 1000;
    getline(&line, &bufsize, arq);

    line = strndup(line, indexOf("films", line) + 10);

    char *aux = (char *)malloc(1000 * sizeof(char));
    char *aux2 = (char *)malloc(TAM * sizeof(char));
    char *aux3 = (char *)malloc(TAM * sizeof(char));

    strcpy(personagem.nome, strndup(&line[10], indexOf("'height':", line) - 13));

    strcpy(aux, strndup(&line[indexOf("'height':", line) + 11], indexOf("'mass':", &line[indexOf("'height':", line) + 11]) - 3));
    if (aux[0] == 'u')
        personagem.altura = 0;
    else
        sscanf(aux, "%d", &personagem.altura);

    strcpy(aux, strndup(&line[indexOf("'mass':", line) + 9], indexOf("'hair_color':", &line[indexOf("'mass':", line) + 9]) - 3));

    int tam = strlen(aux);

    for (int i = 0; i < tam; i++)
    {
        if (aux[i] == ',')
        {
            aux2 = strndup(aux, i);
            strcpy(aux3, &aux[i + 1]);
            strcat(aux2, aux3);
            strcpy(aux, aux2);
        }
    }

    if (aux[0] == 'u')
        personagem.peso = 0;
    else
        sscanf(aux, "%lf", &personagem.peso);
    aux = NULL;
    free(aux);
    aux2 = NULL;
    free(aux2);
    aux3 = NULL;
    free(aux3);

    strcpy(personagem.corDoCabelo, lerCorDoCabelo(line));

    strcpy(personagem.corDaPele, lerCorDaPele(line));

    strcpy(personagem.corDosOlhos, lerCorDosOlhos(line));

    strcpy(personagem.anoNascimento, lerAnoNascimento(line));

    strcpy(personagem.genero, lerGenero(line));

    strcpy(personagem.homeworld, lerHomeworld(line));

    /*printf(" ## %s ## ",personagem.nome);
    printf("%d ## ",personagem.altura);
    printf("%g ## ",personagem.peso);
    printf("%s ## ",personagem.corDoCabelo);
    printf("%s ## ",personagem.corDaPele);
    printf("%s ## ",personagem.corDosOlhos);
    printf("%s ## ",personagem.anoNascimento);
    printf("%s ## ",personagem.genero);
    printf("%s ## \n",personagem.homeworld);*/
}

void printPersonagem(Personagens x)
{
    personagem = x;
    lerPersonagem();

    printf(" ## %s ## ", personagem.nome);
    printf("%d ## ", personagem.altura);
    printf("%g ## ", personagem.peso);
    printf("%s ## ", personagem.corDoCabelo);
    printf("%s ## ", personagem.corDaPele);
    printf("%s ## ", personagem.corDosOlhos);
    printf("%s ## ", personagem.anoNascimento);
    printf("%s ## ", personagem.genero);
    printf("%s ## \n", personagem.homeworld);
}

char *lerCorDoCabelo(char *line)
{
    char *aux = (char *)malloc(1000 * sizeof(char));

    strcpy(aux, strndup(&line[indexOf("'hair_color':", line) + 15], indexOf("'skin_color':", &line[indexOf("'hair_color':", line) + 15]) - 3));

    return aux;
}

char *lerCorDaPele(char *line)
{
    char *aux = (char *)malloc(1000 * sizeof(char));

    strcpy(aux, strndup(&line[indexOf("'skin_color':", line) + 15], indexOf("'eye_color':", &line[indexOf("'skin_color':", line) + 15]) - 3));

    return aux;
}

char *lerCorDosOlhos(char *line)
{
    char *aux = (char *)malloc(1000 * sizeof(char));

    strcpy(aux, strndup(&line[indexOf("'eye_color':", line) + 14], indexOf("'birth_year':", &line[indexOf("'eye_color':", line) + 14]) - 3));

    return aux;
}

char *lerAnoNascimento(char *line)
{
    char *aux = (char *)malloc(1000 * sizeof(char));

    strcpy(aux, strndup(&line[indexOf("'birth_year':", line) + 15], indexOf("'gender':", &line[indexOf("'birth_year':", line) + 15]) - 3));

    return aux;
}

char *lerGenero(char *line)
{
    char *aux = (char *)malloc(1000 * sizeof(char));

    strcpy(aux, strndup(&line[indexOf("'gender':", line) + 11], indexOf("'homeworld':", &line[indexOf("'gender':", line) + 11]) - 3));

    return aux;
}

char *lerHomeworld(char *line)
{
    char *aux = (char *)malloc(1000 * sizeof(char));

    strcpy(aux, strndup(&line[indexOf("'homeworld':", line) + 14], indexOf("'films':", &line[indexOf("'homeworld':", line) + 14]) - 3));

    return aux;
}

char *substring(char *padrao, char *entrada)
{
    char *pointer = strstr(entrada, padrao);

    return strdup(pointer);
}

long int indexOf(char *padrao, char *entrada)
{
    char *pointer = strstr(entrada, padrao);

    return pointer - entrada;
}

bool isFim(char *s)
{
    return (strlen(s) >= 3 && s[0] == 'F' && s[1] == 'I' && s[2] == 'M'); //retornar true ou false para se eh o FIM
}

void inserirFim(Personagens x)
{
    if (n == 0)
    {
        primeiro = novaCelulaDupla(x);
        ultimo = primeiro;
    }

    else
    {
        ultimo->prox = novaCelulaDupla(x);
        ultimo->prox->ant = ultimo;
        ultimo = ultimo->prox;
    }
    n++;
}

File: Questao 07 - Check/_QuickSort.java

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
//import java.io.File;
//import java.util.Scanner;

class Characters {

    // Declaracao de elementos privados
    private String fileName;
    private String name;
    private String height;
    private String weight;
    private String hairColor;
    private String skinColor;
    private String eyeColor;
    private String birthYear;
    private String genre;
    private String homeWorld;

    // Sets and Gets
    public String getFileName() {
        return this.fileName;
    }

    public void setFileName(String fileName) {
        this.fileName = fileName;
    }

    public String getName() {
        return this.name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getHeight() {
        return this.height;
    }

    public void setHeight(String height) {
        this.height = height;
    }

    public String getWeight() {
        return this.weight;
    }

    public void setWeight(String weight) {
        this.weight = weight;
    }

    public String getHairColor() {
        return this.hairColor;
    }

    public void setHairColor(String hairColor) {
        this.hairColor = hairColor;
    }

    public String getSkinColor() {
        return this.skinColor;
    }

    public void setSkinColor(String skinColor) {
        this.skinColor = skinColor;
    }

    public String getEyeColor() {
        return this.eyeColor;
    }

    public void setEyeColor(String eyeColor) {
        this.eyeColor = eyeColor;
    }

    public String getBirthYear() {
        return this.birthYear;
    }

    public void setBirthYear(String birthYear) {
        this.birthYear = birthYear;
    }

    public String getGenre() {
        return this.genre;
    }

    public void setGenre(String genre) {
        this.genre = genre;
    }

    public String getHomeWorld() {
        return this.homeWorld;
    }

    public void setHomeWorld(String homeWorld) {
        this.homeWorld = homeWorld;
    }

    Characters() {
        this.fileName = "";
        this.name = "";
        this.height = "";
        this.weight = "";
        this.skinColor = "";
        this.hairColor = "";
        this.eyeColor = "";
        this.birthYear = "";
        this.genre = "";
        this.homeWorld = "";
    }

    // Ler arquivo txt
    public void readCharacter(String fileName) {
        try {
            FileReader arq = new FileReader(fileName);
            BufferedReader readFile = new BufferedReader(arq);
            String line = readFile.readLine();

            // MyIO.println(line);
            // Chamada de funcoes para separar dados
            nameSetUp(line);
            heightSetUp(line);
            massSetUp(line);
            hair_colorSetUp(line);
            skin_colorSetUp(line);
            eye_colorSetUp(line);
            birth_yearSetUp(line);
            genderSetUp(line);
            homeworldSetUp(line);
            readFile.close();
            arq.close();
        } catch (IOException e) {
        }
    }

    // Separar nome
    public void nameSetUp(String line) {

        int nameBegin = line.indexOf("'name'");
        int nameEnd = line.indexOf("'height'", nameBegin);
        String name = line.substring(nameBegin + 7, nameEnd).trim().replace(",", "").replace("'", "");

        this.name = name;
        // MyIO.println("##" + name);
    }

    // Separar altura
    public void heightSetUp(String line) {

        int heightBegin = line.indexOf("'height'");
        int heightEnd = line.indexOf("'mass'", heightBegin);
        String height = line.substring(heightBegin + 10, heightEnd).trim().replace(",", "").replace("'", "")
                .replace("unknown", "0");

        this.height = height;
        // MyIO.println("##" + height);
    }

    // Separar peso
    public void massSetUp(String line) {

        int massBegin = line.indexOf("'mass'");
        int massEnd = line.indexOf("'hair_color'", massBegin);
        String mass = line.substring(massBegin + 8, massEnd).trim().replace(",", "").replace("'", "").replace("unknown",
                "0");

        this.weight = mass;
        // MyIO.println("##" + mass);
    }

    // Separar cor do cabelo
    public void hair_colorSetUp(String line) {

        int hair_colorBegin = line.indexOf("'hair_color'");
        int hair_colorEnd = line.indexOf("'skin_color'", hair_colorBegin);
        String hair_color = line.substring(hair_colorBegin + 14, hair_colorEnd).trim().replace(",", "").replace("'",
                "");

        this.hairColor = hair_color;
        // MyIO.println("##" + hair_color);
    }

    // Separar cor da pele
    public void skin_colorSetUp(String line) {

        int skin_colorBegin = line.indexOf("'skin_color'");
        int skin_colorEnd = line.indexOf("'eye_color'", skin_colorBegin);
        String skin_color = line.substring(skin_colorBegin + 15, skin_colorEnd).trim().replace(",", "").replace("'",
                "");

        this.skinColor = skin_color;
        // MyIO.println("##" + skin_color);
    }

    // Separar cor do olho
    public void eye_colorSetUp(String line) {

        int eye_colorBegin = line.indexOf("'eye_color'");
        int eye_colorEnd = line.indexOf("'birth_year'", eye_colorBegin);
        String eye_color = line.substring(eye_colorBegin + 14, eye_colorEnd).trim().replace(",", "").replace("'", "");

        this.eyeColor = eye_color;
        // MyIO.println("##" + eye_color);
    }

    // Separar ano de nascimento
    public void birth_yearSetUp(String line) {

        int birth_yearBegin = line.indexOf("'birth_year'");
        int birth_yearEnd = line.indexOf("'gender'", birth_yearBegin);
        String birth_year = line.substring(birth_yearBegin + 14, birth_yearEnd).trim().replace(",", "").replace("'",
                "");

        this.birthYear = birth_year;
        // MyIO.println("##" + birth_year);
    }

    // Separar genero
    public void genderSetUp(String line) {

        int genderBegin = line.indexOf("'gender'");
        int genderEnd = line.indexOf("'homeworld'", genderBegin);
        String gender = line.substring(genderBegin + 11, genderEnd).trim().replace(",", "").replace("'", "");

        this.genre = gender;
        // MyIO.println("##" + gender);
    }

    // Separar planeta natal
    public void homeworldSetUp(String line) {

        int homeworldBegin = line.indexOf("'homeworld'");
        int homeworldEnd = line.indexOf("'films'", homeworldBegin);
        String homeworld = line.substring(homeworldBegin + 14, homeworldEnd).trim().replace(",", "").replace("'", "");

        this.homeWorld = homeworld;
        // MyIO.println("##" + homeworld);
    }
}

class CelulaDupla {
    public Characters elemento; // Elemento inserido na celula.
    public CelulaDupla ant, prox; // Aponta a celula prox.

    public CelulaDupla() {
        // this.elemento = new Characters();
        this.ant = this.prox = null;
    }

    public CelulaDupla(Characters elemento) {
        this.elemento = elemento;
        this.ant = this.prox = null;
    }
}

class ListaSequencial {

    private CelulaDupla primeiro;
    private CelulaDupla ultimo;
    // private Characters personagem[];
    // private int contador;

    /**
     * Construtor da classe que cria uma lista sem elementos (somente no cabeca).
     */
    public ListaSequencial() {
        primeiro = new CelulaDupla();
        ultimo = primeiro;
        // personagem = new Characters[1000];
        // contador = 0;
    }

    // Insere elemento na ultima posicao da lista
    public void inserirFinal(Characters jedi) {
        ultimo.prox = new CelulaDupla(jedi);
        ultimo.prox.ant = ultimo;
        ultimo = ultimo.prox;
    }

    public void quickSort() {
        quickSort(primeiro.prox, ultimo);
    }

    private void quickSort(CelulaDupla esq, CelulaDupla dir) {
        CelulaDupla temp1 = esq;
        CelulaDupla temp2 = dir;
        int j = posicao(esq);
        int k = posicao(dir);
        CelulaDupla pivo = primeiro.prox;

        for (int i = 0; i < (j + k) / 2; i++, pivo = pivo.prox)
            ;

        while (j <= k) {
            while (compareTo(temp1.elemento.getHairColor(), pivo.elemento.getHairColor(), temp1.elemento.getName(),
                    pivo.elemento.getName()) < 0) {
                temp1 = temp1.prox;
                j++;
            }

            while (compareTo(temp2.elemento.getHairColor(), pivo.elemento.getHairColor(), temp2.elemento.getName(),
                    pivo.elemento.getName()) > 0) {
                temp2 = temp2.ant;
                k--;
            }

            if (j <= k) {
                swap(getCelulaDupla(j), getCelulaDupla(k));

                j++;
                k--;
            }

            if (posicao(esq) < k) {
                quickSort(esq, getCelulaDupla(k));
            }

            if (j < posicao(dir)) {
                quickSort(getCelulaDupla(j), dir);
            }
        }
    }

    int compareTo(String cabelo1, String cabelo2, String nome1, String nome2) {

        if (cabelo1.compareTo(cabelo2) == 0) {
            return nome1.compareTo(nome2);
        } else {
            return cabelo1.compareTo(cabelo2);
        }
    }

    public int posicao(CelulaDupla celula) {
        int pos = 0;

        for (CelulaDupla i = primeiro.prox; i != celula; i = i.prox, pos++)
            ;

        return pos;
    }

    public CelulaDupla getCelulaDupla(int pos) {
        CelulaDupla celula = primeiro.prox;

        for (int i = 0; i < pos; i++, celula = celula.prox)
            ;

        return celula;
    }

    public void swap(CelulaDupla celula1, CelulaDupla celula2) {
        CelulaDupla aux = new CelulaDupla(celula1.elemento);

        celula1.elemento = celula2.elemento;
        celula2.elemento = aux.elemento;
        aux = null;
    }

    public void mostrar() {
        int j = 0;
        // System.out.print("[ ");
        for (CelulaDupla i = primeiro.prox.prox.prox.prox; i != null; i = i.prox, j++) {
            // MyIO.print(i.elemento.getName() + " ");
            // MyIO.print("[" + j + "] ## ");
            MyIO.print(" ## " + i.elemento.getName() + " ## ");
            MyIO.print(i.elemento.getHeight() + " ## ");
            MyIO.print(i.elemento.getWeight() + " ## ");
            MyIO.print(i.elemento.getHairColor() + " ## ");
            MyIO.print(i.elemento.getSkinColor() + " ## ");
            MyIO.print(i.elemento.getEyeColor() + " ## ");
            MyIO.print(i.elemento.getBirthYear() + " ## ");
            MyIO.print(i.elemento.getGenre() + " ## ");
            MyIO.println(i.elemento.getHomeWorld() + " ## ");
        }
        // System.out.println("] ");
    }

    public boolean pesquisar(Characters jedi) {
        boolean resp = false;
        for (CelulaDupla i = primeiro.prox; i != null; i = i.prox) {
            if (i.elemento == jedi) {
                resp = true;
                i = ultimo;
            }
        }
        return resp;
    }

    public int tamanho() {
        int tamanho = 0;
        for (CelulaDupla i = primeiro; i != ultimo; i = i.prox, tamanho++)
            ;
        return tamanho;
    }

}

class _QuickSort {
    // Identificar palavra FIM
    public static boolean isFIM(String input) {
        return (input.length() >= 3 && input.charAt(0) == 'F' && input.charAt(1) == 'I' && input.charAt(2) == 'M');
    }

    public static String ISO88591toUTF8(String strISO) {
        try {
            byte[] isoBytes = strISO.getBytes("ISO-8859-1");
            return new String(isoBytes, "UTF-8");
        } catch (IOException e) {
        }
        return strISO;
    }

    public static void main(String[] args) throws Exception {
        // Inicializacao de lista
        ListaSequencial list = new ListaSequencial();
        // Array de characters
        Characters[] character = new Characters[100];

        String[] input = new String[700];
        int enterNumber = 0;

         MyIO.setCharset("UTF-8");

        // Leitura de dados
        do {
            input[enterNumber] = MyIO.readLine();
            input[enterNumber] = ISO88591toUTF8(input[enterNumber]);
        } while (isFIM(input[enterNumber++]) == false);
        enterNumber--;

        for (int i = 0; i < enterNumber; i++) {
            character[i] = new Characters();
            character[i].readCharacter(input[i]);
            list.inserirFinal(character[i]);
        }

        list.quickSort();
        list.mostrar();
    }
}
